### PR TITLE
ENH: Enable accelerated linear regression for overdetermined systems

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

As a follow-up from oneDAL PR: https://github.com/oneapi-src/oneDAL/pull/2930

The PR above enables fitting linear regressions to overdetermined systems, both in regular regression and multi-output regression. Currently, the `sklearnex` estimators fall back to `scikit-learn` for overdetermined systems:
https://github.com/intel/scikit-learn-intelex/blob/2fccf444f800fb8ae1cce844b35727efc623e6d3/sklearnex/linear_model/linear.py#L180

But after the PR above, those should be supported in oneDAL too. Hence, this PR enables using oneDAL for those situations and adds tests for them.

One thing to note though: the mechanism in oneDAL is implemented as a fallback from Cholesky - that is, it will first perform a Cholesky factorization, and if that fails, will then use the fallback solver which can handle overdetermined systems. But if e.g. there are more columns that rows, we'll already know beforehand that Cholesky will invariably fail, and the procedure could be made faster if such kind of solver could be requested straight away from oneDAL. Implementation of it as a selectable solver is left for future PRs.

Note that since the oneDAL PR hasn't been merged at the time of writing, this PR will of course end up failing tests, but will keep it here in the meantime.

---

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by icfaust** [comment]
Due to the nature of using two repos together, you'll unfortunately need to if statement out the new functionality with a ```daal_check_version``` https://github.com/intel/scikit-learn-intelex/blob/main/daal4py/sklearn/_utils.py#L69  Once the onedal PR is merged, we can force run the nightly and then the GitHub Actions CI will reflect the change in oneDAL linear Regression, but the Azure Pipelines will have to wait to a minor release (i.e. 2025.1). This should solve the segfaulting in at least the azure pipelines-side of CI (and again apologies as this is undocumented convention).

**Comment by david-cortes-intel** [comment]
> Due to the nature of using two repos together, you'll unfortunately need to if statement out the new functionality with a `daal_check_version` https://github.com/intel/scikit-learn-intelex/blob/main/daal4py/sklearn/_utils.py#L69 Once the onedal PR is merged, we can force run the nightly and then the GitHub Actions CI will reflect the change in oneDAL linear Regression, but the Azure Pipelines will have to wait to a minor release (i.e. 2025.1). This should solve the segfaulting in at least the azure pipelines-side of CI (and again apologies as this is undocumented convention).

Thanks, added a check for the oneDAL version that would enable features and tests accordingly.

**Comment by david-cortes-intel** [comment]
PR adding the functionality in oneDAL was merged, so this PR should start working by tomorrow as the builds finish I guess.

**Comment by david-cortes-intel** [comment]
CI jobs are passing now.

File: doc/sources/algorithms.rst
Comment (Vika-F) [inline doc/sources/algorithms.rst:307]: In my opinion, it's better to make this statement more specific:
```suggestion
     - Only dense data is supported, `#observations` should be >= `#features` and there should be no linearly dependent features in the data.
```

File: onedal/linear_model/tests/test_linear_regression.py
Comment (Vika-F) [inline onedal/linear_model/tests/test_linear_regression.py:162]: `multivariate_normal.rvs` from `scipy` might be more suitable in this case

File: sklearnex/linear_model/linear.py
Comment (Vika-F) [inline sklearnex/linear_model/linear.py:167]: Please be more specific, i.e. "Multiple output regression was introduced in oneDAL 2025.1", etc.
Maybe also makes sense to move this comment closer to the checks.

**Review by Vika-F — COMMENTED** [review]
The changes look good to me, there are only couple of minor suggestions.

File: onedal/linear_model/tests/test_linear_regression.py
Comment (david-cortes-intel) [reply onedal/linear_model/tests/test_linear_regression.py:162]: You could say that it'd make a more reasonable target to want to run multiple regression on, but from the test's perspective, that shouldn't matter as long as the LHS matrix (which doesn't use the response variables) is not positive-definite.

**Review by samir-nasibli — COMMENTED** [review]
Some comments and suggestions from my side

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:190]:
```suggestion
                    not is_multi_output or supports_all_variants,

```
I think?

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:185]:
```suggestion
                    not is_underdetermined or (supports_all_variants and not is_gpu)),

```
I was getting confused by the logic of the double negations.

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:169]: https://github.com/intel/scikit-learn-intelex/blob/903fb2f97999d6e4acee02c771311a275412294a/onedal/utils/validation.py#L360  Try to use onedal's _num_features, it may be more robust.

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:242]: While this reads like a regression in capability, maybe I am missing part of the picture. Was multi-output previously supported, or only on underdetermined systems? @Vika-F

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:242]: It was technically supported, but if the Cholesky factorization or solve failed, it would either throw an error, or segfault, or corrupt memory, depending on how the PLU factorization goes.

File: sklearnex/linear_model/linear.py
Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:169]: Looks like it doesn't work with lists:
```
TypeError: Unable to find the number of features from X of type list
```

Comment (icfaust) [reply sklearnex/linear_model/linear.py:169]: Could you try instead with sklearn's _num_features instead?

File: sklearnex/linear_model/linear.py
Comment (icfaust) [reply sklearnex/linear_model/linear.py:242]: Gotcha, which is why there is the sklearn_after_onedal on RuntimeError. As of now this would block using it in the case that the Cholesky succeeds (which may also occur).  Since this is a bit of a gray zone, I would want the weigh-in of @napetrov or @syakov-intel .

File: sklearnex/linear_model/linear.py
Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:169]: Same error.

Comment (icfaust) [reply sklearnex/linear_model/linear.py:169]: Could you provide the list data which is causing _num_features to TypeError? 
I assume you are doing
```python
is_multi_output = _num_features(y) > 1
```

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:169]: Example where it errors passing a list:
```python
from sklearn.utils.validation import _num_features
_num_features([1,2,3])
```
```
TypeError: Unable to find the number of features from X of type list
```

Comment (icfaust) [reply sklearnex/linear_model/linear.py:169]: Use onedal's with _num_features(X, True) (has a fallback for 1d in the onedal implementation)

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py]: Changed to use oneDAL's `_num_features`.

File: sklearnex/linear_model/linear.py
Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py]: Changed to initialize `PatchingConditionsChain` on the CPU/GPU versions of each function - not sure if this is what you had in mind though.

Comment (icfaust) [reply sklearnex/linear_model/linear.py]: It looks great, thank you for making the change.

**Review by icfaust — APPROVED** [review]
Approval contingent on this conversation being resolved https://github.com/intel/scikit-learn-intelex/pull/2083#discussion_r1808624325 . I believe this change may need to be modified based on previous experience, but they should give a definitive answer.  Otherwise its good to go.

**Comment by david-cortes-intel** [comment]
I'll be playing a bit with the CIs now to see at which commit the jobs broke.

File: sklearnex/linear_model/linear.py
Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py]: Turns out `_num_features` behaves differently for different input types, so it's not a good replacement:
```python
import numpy as np
from onedal.utils import _num_features
_num_features([1,2,3], True) # =1
_num_features(np.array([1,2,3]), True) # =3
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
PreCommit has few repeated failures of LinearRegression. Also, need to rule out all other failures as sporadic.

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:193]: Why is this here?

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:193]: Different behavior or `_num_feature` with lists and arrays:
https://github.com/intel/scikit-learn-intelex/pull/2083#discussion_r1810369965

File: sklearnex/linear_model/linear.py
Comment (icfaust) [inline sklearnex/linear_model/linear.py:19]: Will need to be removed

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py]: Removed.

File: sklearnex/linear_model/linear.py
Comment (icfaust) [reply sklearnex/linear_model/linear.py:192]: https://github.com/intel/scikit-learn-intelex/blob/f350c0d183f3a2503ad9a09dd36180d68cef713b/onedal/utils/validation.py#L379 is definitely the cause.  We can change it in this PR or we can make a small bugfix and run it against CI. Your choice!

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py]: Looks like changing `_num_features` to have the desierd behavior didn't break any other tests, so I'm modifying it here.

**Comment by icfaust** [comment]
@DDJHB please include these style changes into ridge out of preview, should enable some deselected tests.

File: sklearnex/linear_model/linear.py
Comment (napetrov) [reply sklearnex/linear_model/linear.py:249]: I think worth discussing at arch meeting - i don't have preference

File: sklearnex/linear_model/linear.py
Comment (Alexsandruss) [inline sklearnex/linear_model/linear.py:142]:
```suggestion
        patching_status = PatchingConditionsChain(
            f"sklearn.linear_model.{self.__class__.__name__}.{method_name}"
        )
```

File: sklearnex/linear_model/linear.py
Comment (Alexsandruss) [inline sklearnex/linear_model/linear.py:147]:
```suggestion
            f"sklearn.linear_model.{self.__class__.__name__}.{method_name}"
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]
Patching logging got broken due to incorrect names.

**Comment by Alexsandruss** [comment]
http://intel-ci.intel.com/ef96e180-8fcb-f125-9c69-a4bf010d0e2d

**Review by Alexsandruss — APPROVED** [review]

File: sklearnex/linear_model/linear.py
Comment (icfaust) [reply sklearnex/linear_model/linear.py:249]: @david-cortes-intel I would say lets merge this PR after we add a ticket to the backlog to address it.

Comment (david-cortes-intel) [reply sklearnex/linear_model/linear.py:249]: @icfaust To clarify: the issue (potentially crashing multi-output regression) happens in older releases of oneDAL, but is solved in the current GH main branch, which should form part of the 2025.1 release.

**Pull request merged**
